01-data-structures/03-linked-lists/linked-lists-answers.md

```diff
@@ -0,0 +1,47 @@
+# Questions about Linked Lists
+
+## Q1. What is Spatial Locality and why does it benefit performance?
+Spatial locality is when related elements are physically close together in memory. 
+For example, in an array, all the elements of the array are next to each other.
+This makes several operations on data structures fast:
+* Fetching all data associated with a data structure can be done at once, without
+a lot of searching or following pointers.
+* Navigating (such as going to the previous element or next element) is trivial.
+
+## Q2. Compare the performance of an Array to a Linked List using the Benchmark module.
```

Comment:
Overall, understand two caveats with this comparison:

1) The untold secret: Ruby's arrays are "dynamic" versus "static". "Static" arrays are created with a hard size and take up that much memory exactly. The memory they hold is lined up sequentially, so sometimes it is impossible to create a static array of a large size if your system simply does not have that much space available. "Dynamic" arrays similarly are allocated in a contiguous chunk of a given size, but will grow and shrink themselves as they go. In Ruby, they will also fragment themselves if necessary, thus making them behave as a weird half-array-half-linked-list sort of structure. Ruby will determine which of these operations to take when based on run-time analysis of performance implications.

2) The more obvious one: your linked list written in Ruby is competing against industry software engineers' array written in C. And given these arrays do not have to deal with every caveat of static arrays, operations like "delete from middle", which would usually cause items to be shifted in memory, can bleed every bit of performance out of these arrays as computationally possible. It is hardly a fair right.